contracts/LinearReverseDutchAuction.sol

```diff
@@ -0,0 +1,170 @@
+        require(reservedPrice>=0,"Reserved price cannot be negative");
+        require(startingPrice>=0,"Starting price cannot be negative");
+        require(startingPrice>=reservedPrice,"Starting price should be higher than reserved price");
+        require(duration>0,"Duration must be greater than zero seconds");
+        require(auctionedToken!=address(0),"Auctioned token cannot be zero address");
+        require(biddingToken!=address(0),"Bidding token cannot be zero address");
+
+        if(auctionType == AuctionType.NFT){
+            require(IERC721(auctionedToken).ownerOf(auctionedTokenIdOrAmount)==msg.sender,"Caller must be the owner");
+            IERC721(auctionedToken).safeTransferFrom(msg.sender,address(this),auctionedTokenIdOrAmount);
+        }else{
+            require(IERC20(auctionedToken).balanceOf(msg.sender)>=auctionedTokenIdOrAmount,"Insufficient balance");
+            SafeERC20.safeTransferFrom(IERC20(auctionedToken),msg.sender,address(this),auctionedTokenIdOrAmount);
+        }
+        
+        uint256 deadline=block.timestamp+duration;
+
+        auctions[auctionCounter] = AuctionData({
+            id: auctionCounter,
+            name:name,
+            description:description,
+            imgUrl:imgUrl,
+            auctioneer: msg.sender,
+            auctionType: auctionType,
+            auctionedToken: auctionedToken,
+            auctionedTokenIdOrAmount: auctionedTokenIdOrAmount,
+            biddingToken: biddingToken,
+            startingPrice: startingPrice,
+            availableFunds: 0,
+            reservedPrice: reservedPrice,
+            winner: msg.sender,
+            deadline: deadline,
+            duration: duration,
+            isClaimed: false
+        });
+
+        emit AuctionCreated(
+            auctionCounter++,
+            name,
+            description,
+            imgUrl,
+            msg.sender,
+            auctionType,
+            auctionedToken,
+            auctionedTokenIdOrAmount,
+            biddingToken,
+            startingPrice,
+            reservedPrice,
+            deadline
+        );
+    }
+
+    function getCurrentPrice(uint256 auctionId) public view validAuctionId(auctionId) returns (uint256) {
+        AuctionData storage auction=auctions[auctionId];
+        require(block.timestamp<auction.deadline,"Auction has ended");
+        require(!auction.isClaimed,"Auction has ended");
+
+        // price(t) = startingPrice - (((startingPrice - reservedPrice) * (timeElapsed)) / duration)
+        return auction.startingPrice - (((auction.startingPrice-auction.reservedPrice)*(block.timestamp - (auction.deadline - auction.duration)))/auction.duration);
+    }
+
+    //placeBid function is not required in this auction type as the price is determined by the auctioneer and the auctioneer can withdraw the item on placing the bid only
+    function withdrawItem(uint256 auctionId,uint256 bidAmount) external validAuctionId(auctionId) {
```

Comment:
I think this function is the same for all ReverseDutchAuction types. Therefore, we can create an abstract superclass called `ReverseDutchAuction`, a subclass of `Auction`, containing this function. Then the subclasses of `ReverseDutchAuction` will just have to implement `getCurrentPrice`.